std/traits.d

```diff
@@ -7969,15 +7973,15 @@ has both opApply and a range interface.
 */
 template ForeachType(T)
 {
-    alias ForeachType = ReturnType!(typeof(
+    alias ForeachType = typeof(
     (inout int x = 0)
     {
         foreach (elem; T.init)
         {
             return elem;
         }
         assert(0);
-    }));
+    }());
```

Comment:
It's subtle, but this isn't going to work for `inout` either. I found out the hard way.

What happens when `T` is `inout`:

1. The foreach is allowed inside, because `(inout T).init` can be used, simply for the fact that `inout int` is in the parameter list.
2. The way inout works is that it takes on the mutability of the parameter upon return. What is the parameter mutability for `inout int = 0`? mutable.
3. The return value will be *stripped* of `inout`, which means e.g. `ForeachType!(inout(int)[]) == int` instead of `inout int` (like it currently is)

The reason `ReturnType` works is because it constructs an argument matching the type of the parameter, e.g. it declares an `inout int` to pass in.

In this case, I think `(ref T val) {...}(lvalueOf!T)` is the appropriate solution.

Reply:
ugh, I take it all back. What I did before was different, and actually, this does work. Nevermind.